### PR TITLE
Fix for BinaryValue type

### PR DESCRIPTION
**Pull request by Matthewar**
A number of issues have noted an problem with the `BinaryValue` type, which seems to centre around the `big_endian` member variable. It is treated as both byte-order and bit-order in different places in the code, so I have attempted to clarify this in comments/function docstrings.

- When the binary value type was first added, there was no concept of endian-ness ([commit](https://github.com/Matthewar/cocotb/commit/55023b9644aaf0574579edc07e30a1166f166550#diff-77cf4987173ad37eab355c44a547ffefR56))
- The `bigEndian` argument was added to the constructor/`big_endian`member added to the class to convert to/from a bytestring IE it represents byte-ordering ([commit](https://github.com/Matthewar/cocotb/commit/761ead050745328d317250a6c20898aaa78db8cd#diff-77cf4987173ad37eab355c44a547ffefR106))
- When signed representation was added to the class, `big_endian` was used for bit ordering ([commit](https://github.com/Matthewar/cocotb/commit/030a25bdeae4d02db12becf461f33c5d4217c67f))

Big-endian doesn't seem to have been used or tested thoroughly, at least for when it represents bit ordering, which makes sense as hardware designs typically don't use a big endian bit order, and the languages don't really have the concept of bit-endianness in the same way:

- In 1364-2005 (the verilog specification) section 4.3.1: "The most significant bit specified by the _msb_ constant expression is the left-hand value in the range, and the least signficant bit specified by the _lsb_ constant expression is the right-hand value in the range".
    - Implicit extension/truncation will occur on the left side, even if an ascending range (e.g. `logic [0:3]`) is declared.
    - Implicit integer conversion will always result in the vector being treated as having a little-endian bit order
- In the 1076-2002 (VHDL revision) section 16.8.5.1 which defines the `numeric_std`package, `signed` and `unsigned` types clearly define: "The type UNSIGNED represents an unsigned binary integer with the most significant bit on the left, while the type SIGNED represents a two's-complement binary integer with the most significant bit on the left."
    - Explicit extension/truncation can occur on whichever side (left/right) you want, so for big-endian bit order signals (e.g. `signal example : std_logic_vector(0 to 9);`) you can extend/truncate on the right side if you'd prefer
    - The `RESIZE` function acts like Verilog implicit truncation/extension and always operations on the left side
    - `TO_INTEGER` function takes a little-endian bit-order vector

To solve this:
- Returned `big_endian` member to represent the byte order of the internal string
    - When calling `get_buff`/`set_buff`, it is assumed you are using a big-endian byte order value, so byteswapping will only occur if the internal value is being treated as little-endian byte order
    - Wasn't sure if this was the best way to go about this, but `bigEndian=True` has been the default behaviour for a long time, even though most verilog/vhdl signals aren't big-endian byte order (otherwise the bit indices would go `7,6,5,4,3,2,1,0,15,14,13,12,11,10,9,8`, I think?)
    - Haven't done this yet, but maybe should change `handle.py` to return `BinaryValue`s with `bigEndian=False`
- Added `ascending_range` member
    - This is the bit-ordering (`ascending_range=True` => big-endian bit order)
    - Used for conversion between integer forms and string buffer
    - Used for indexing the bit string
        - Only when you index `BinaryValue`, not when you index the result of `get_binstr`, which returns a Python string, which will always have the leftmost value at index zero
        - Don't flip the binary string outputted so it's indexable correctly, it's showing directly what is in the simulation

Also:
- Added simulations to show the behaviour of bit-endianness in Verilog/VHDL, confirm behaviour of `BinaryValue.ascending_range`
- Added check to avalon drivers/monitors to ensure `dataBitsPerSymbol` is always 8, as that means `firstSymbolInHigherOrderBits` represents the byte order, which is how it has been used previously
- Removed `other` argument from `__neg__`, which doesn't take arguments
- Made `modulo` argument for `__ipow__` method optional as defined
- __Deprecated__ `signed_integer` member/getter
    - Added in [commit](https://github.com/Matthewar/cocotb/commit/f3008e5478d6b806552b6c664f65e28ca97086ac)
    - Before signed representation/endian-ness were added
    - Integer member/setter/getter handles signed values anyway, this isn't needed and just seems to force the value to act signed using a different method to the other code
- Some formatting changes - missing docstrings, some incorrect indentation, comments, etc.

### CODE REVIEW

File: tests/test_cases/test_array/test_array.py [hunk 1]
```diff
@@ -326,22 +326,22 @@ def test_discover_all(dut):
     elif cocotb.LANGUAGE in ["verilog"] and cocotb.SIM_NAME.lower().startswith(("riviera")):
         if cocotb.SIM_VERSION.startswith(("2017.10.61")):
             pass_total = 803
-        elif cocotb.SIM_VERSION.startswith(("2016.06","2016.10","2017.02")):
+        elif cocotb.SIM_VERSION.startswith(("2016.06", "2016.10", "2017.02")):
```
Comment (eric-wieser) [inline tests/test_cases/test_array/test_array.py]: This is good whitespace cleanup, but this PR is not the place for it. Can you move these trivial style changes to a separate PR, so that it makes the non-trivial content of this one easier to review?

Comment (Matthewar) [reply tests/test_cases/test_array/test_array.py]: Sure thing, I just did some of them so my linter would stop complaining at me!

**Comment by mciepluc** [comment]
@Matthewar I have started similar work in #706 specifically to fix #608 . As this PR seems to address much more stuff, seems to be better to focus on this one. I however recommend for you to include the test case from #706 that I have prepared (tests/test_cases/issue_608) as it may catch some inconsistencies.

**Comment by Matthewar** [comment]
Ah, apologies, I think I missed that. Sure, I will merge that in, and can also copy in some of the details from that PR message as it is also quite detailed.

File: tests/test_cases/test_array/test_array.py (hunk 1 above)
Comment (Matthewar) [reply tests/test_cases/test_array/test_array.py]: All the style changes, or just ones in standalone files?

Comment (Matthewar) [reply tests/test_cases/test_array/test_array.py]: https://github.com/potentialventures/cocotb/pull/1022

Once that's merged to master, I'll merge it here and it should remove a lot of the whitespace changes.

**Comment by Matthewar** [comment]
From #608:

https://github.com/potentialventures/cocotb/issues/608#issuecomment-379543071 + results tested with the changes from this PR.
> init int | init length | init bigEndian | .binstr | .integer | new with fix |
> -- | -- | -- | -- | -- | -- |
> 3 | 3 | True | 110 | 6 | 3
> 4 | 3 | True | 100 | 4 | 4
> 3 | 3 | False | 011 | 3 | 3
> 4 | 3 | False | 100 | 4 | 4
> 5 | 5 | True | 10100 | 20 | 5
> 2 | 5 | True | 10000 | 16 | 2
> 5 | 5 | False | 00101 | 5 | 5
> 2 | 5 | False | 00010 | 2 | 2

https://github.com/potentialventures/cocotb/issues/608#issuecomment-409080667 but with this PR.
> ```
> >>> BinaryValue(12).integer
> 12
> >>> BinaryValue(12, bigEndian=False).integer
> 12
> >>> BinaryValue(12, bigEndian=False, bits=8).integer
> 12
> >>> BinaryValue(12, bits=8).integer
> 12
> ```

Note that `bigEndian` here means byte order so actually has nothing to do with bit-ordering, and `ascending_range` defaults to `False` IE little-endian bit order. The only difference in results will be seen in the `binstr`/`buff`, e.g.
```python
>>> BinaryValue(value=3, n_bits=3, ascending_range=True).binstr
'110'
>>> BinaryValue(value=3, n_bits=3, ascending_range=False).binstr
'011'
```

**Comment by imphil** [comment]
Thanks @Matthewar for tackling the problems with BinaryValue, that's a longstanding pain and very much appreciated. Can you rebase this to remove the whitespace-only changes and make review easier? Thanks!

**Comment by Matthewar** [comment]
I've merged in master, I can rebase it if you'd prefer, please let me know. Also, one of the merge conflicts related to this commit, please confirm:
- For commit 9a0039c422981da7c06052b788b785eae3dc2dbf
- It removes `modulo` argument from `__pow__` and `__ipow__`
- But the Python data model docs show it as an optional https://docs.python.org/3.7/reference/datamodel.html?highlight=__ipow__#object.__ipow__
- I had changed them from required to optional

Is this fine or is there a reason they should be removed?

**Comment by imphil** [comment]
A rebase would be very much appreciated, leading to a clean series of commits only containing changes to improve BinaryValue.

Your question on the `modulo` argument removal in https://github.com/potentialventures/cocotb/commit/9a0039c422981da7c06052b788b785eae3dc2dbf: @eric-wieser  can you jump in here?

**Comment by Matthewar** [comment]
Will rebase once I know what to do with `modulo`

**Comment by Matthewar** [comment]
I'm not sure if I'm happy with how `big_endian`/`bigEndian` is used (including in this PR), this is something that could be considered.
- Presumably, `bigEndian=True` (byte-order) was a default as network byte-order is big endian
    - This means `set_buff`/`get_buff` methods can be easily used with packages such as scapy
    - They're expected to take/provide a big-endian (byte order) byte-string
        - Maybe we should add an option to specify what byte-order you want the string in in `get_buff`/`set_buff`, as an aside, to make things clearer
- As far as I understand, Verilog/VHDL single dimensional vectors don't really care about byte-order endian-ness
    - As I detailed in the top comment, they don't really care about bit-order endian-ness either, left = top in both languages
    - If we were to care about byte-ordering here, this is how I would think the bit indices would look
        - `Bits` here is going to represent the bit positions of the underlying `binstr`

```
 big_endian=True; ascending_range=True (big endian bit/byte order)
Bits:  0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15
Bytes: | ---- 0 ---- | | ------- 1 ------- |

big_endian=False; ascending_range=False (little endian bit/byte order)
Bits:  15 14 13 12 11 10 9 8 7 6 5 4 3 2 1 0
Bytes: | ------- 1 ------- | | ---- 0 ---- |

big_endian=True; ascending_range=False (big endian byte order, little endian bit order)
Bits:  7 6 5 4 3 2 1 0 15 14 13 12 11 10 9 8
Bytes: | ---- 0 ---- | | ------- 1 ------- |

big_endian=False; ascending_range=True (little endian byte order, big endian bit order)
Bits:  8 9 10 11 12 13 14 15 0 1 2 3 4 5 6 7
Bytes: | ------- 1 ------- | | ---- 0 ---- |
```
- To me, only two of those forms really make sense when you're considering single dimensional vectors in hardware, when `big_endian == ascending_range`.
- When considering software, either form where `ascending_range=False` looks sensible, as within a byte my understanding is that bits are in little-endian order
- If you can flatten 2D vectors to a `BinaryValue` (maybe you can through `simulator.get_signal_val_binstr`, I'm not sure), then the other forms may also make sense, if you had a Verilog signal `logic [0:3][7:0] signal_2d;` for example

Currently, `big_endian` is completely ignored when you index `BinaryValue` objects (or convert them to integers), which could be considered to be the wrong behaviour, e.g. for a `big_endian=True; ascending_range=False` value theoretically has its most significant bit in the middle, but will be indexed as if bit 15 is at the start of the binstr.
- Could use `big_endian` in calculations for integer conversions and indexing
    - Quite confusing (just look at the 4 different scenarios)
    - In hardware only two will be used, should probably change default in `handle.py` to return `big_endian=False` too in this case
- Could keep it the way it is and slightly change the description/meaning behind `big_endian`
    - Not sure what we would change it to, but need to make it clear when `binstr` has a different byte order to `buff` (`binstr` show underlying bits, `buff` shows bytes in big-endian byte order)
- Could have this as two classes, one that represents byte order and one that represents the underlying bit order value
    - Hard to do this without making a breaking change, which should be avoided
    - Adds complexity to the software stack

I'm not sure what the best solution for this is - there are probably other ways to resolve this, please free free to suggest them. I think at the moment I'm leaning towards the first, using both members in integer conversions and indexing, and changing `handle.py` to default return `big_endian=False` when calling `dut.signal.value` (which is does in at least one place anyway).
I don't think we can change the constructor default to `bigEndian=False` because that'll break functionality (maybe we can, I'm not sure how people used it at all, looking at how broken the code was), but the result of this choice could lead to some confusion, as it would affect binary strings when assigning values:
```python
>>> BinaryValue(value=287, bigEndian=False, n_bits=16).binstr
'0000000100011111'
>>> BinaryValue(value=287, bigEndian=True, n_bits=16).binstr
'0001111100000001'
```

**Comment by nevsan** [comment]
At the very least, I would love to see the name change from `bigEndian` to `ascending` or some other name that doesn't conflict with a term that should only be used to describe byte ordering (endianness). I was really happy when I saw this PR just to have that fix alone.

**Comment by eric-wieser** [comment]
I think this type of change is going to break a lot of people's tests, even if overall it is a good thing. Do we care about that? Can we make the new semantics opt in somehow?

**Comment by Matthewar** [comment]
In its current form it shouldn't "break" things, as they were relying on broken behaviour (unless it could be argued that `big_endian` can mean both bit and byte order at the same time, but that seems messy).

Making `big_endian` mean both bit and byte order could be another solution rather than the ones I mentioned in https://github.com/cocotb/cocotb/pull/1020#issuecomment-511798563
- Could mark `big_endian` for deprecation and say it means bit-order and byte-order
    - Scenarios `big_endian=ascending_range` in the above comment
    - However if you get a bytestring `get_buff`, you'd be getting a bytestring of big-endian bit order values
        - not the normal behaviour of network byte-order
        - Usually little-endian bit order within a byte
        - Could change `get_buff` to flip this but then would `set_buff` have different behaviour? Not necessarily consistent
- Then have `ascending_range` as the new form of `big_endian` (like we've done with n_bits and bits)
- Create wrapper for `BinaryValue` to handle byte-ordering on top of the vectors

Not sure if that's any cleaner/less break-y than the other approaches though.

**Comment by eric-wieser** [comment]
> as they were relying on broken behaviour

Sometimes I've found the broken behavior is forced upon me, so I haven't really had a choice but to try and work around it. The danger is that workarounds people put in place because they did not understand the broken behavior are likely to now make things worse.

**Comment by eric-wieser** [comment]
> Maybe we should add an option to specify what byte-order you want the string in in get_buff/set_buff

+1 from me, that will make it match the builtin `int.to_bytes(byteorder='big')`. Personally I'd liked to see `get_buff` and `set_buff` deprecated, and replaced with `some_value.to_bytes(byteorder=...)` and `some_value = BinaryValue.from_bytes(..., n_bits=...)`

**Comment by eric-wieser** [comment]
Here's one route a deprecation could go:
1. Create a new `BinaryValue2` class, for want of a better name, that has all of the fixes we could dream of
2. Change `Handle.value` to return a `WarningBinaryValue`, which emits a warning upon forwarding attribute access with the message _"In future, `.value` will return a `BinaryValue2` instance. To opt into this early, use `BinaryValue2(handle.value)`. To preserve the old behavior, use `CompatBinaryValue(handle.value)`"_
3. In future, change  `Handle.value` to return `BinaryValue2`

**Comment by Matthewar** [comment]
Alright, this has been rebased - but I haven't yet handled the `big_endian` discussion we've been having here. I feel like the best way to change this without breaking things at this point is to use `bigEndian` as bit-order, which essentially means it acts as both bit and byte order (although in byte-order terms, the bit-order is the wrong way round in each byte). We can mark the `bigEndian` argument as deprecated and add `ascending_range`, just because it's a bit confusing.
Then change the `set_buff`/`get_buff` to reverse the byte-order when `bigEndian=False`, and flip the bit order within bytes when `bigEndian=True` IE return/accept the expected network-order byte string to those functions.

I'm happy to also make another version of `BinaryValue` that acts more cleanly and can import the old one to "upgrade" it, but I'll do that in another issue.

File: cocotb/binary.py [hunk 2]
```diff
@@ -91,29 +96,42 @@ class BinaryValue(object):
...
+        # - This is a Python string of bits
+        # - Note that because it is a string, if acending_range is false, indexing (of strings) no
+        #   longer provides the expected result (0th index is still the start of the string)
         self._str = ""
+
         self.big_endian = bigEndian
         self.binaryRepresentation = binaryRepresentation
+        self.ascending_range = ascending_range
+
+        self._log = SimLog("cocotb.binary.%s" % self.__class__.__name__)
```
Comment (eric-wieser) [inline cocotb/binary.py]: Having a log lookup per binary value might be massively expensive. Can you make this a class-level member?

Comment (Matthewar) [reply cocotb/binary.py]: Is that as simple as moving it out of the constructor as `_log = SimLog("cocotb.binary.%s" % BinaryValue.__class__.__name__)` ?

Comment (eric-wieser) [reply cocotb/binary.py]: Yes - but `BinaryValue.__class__.__name__` is a weird way to spell `"BinaryValue"`

Comment (Matthewar) [reply cocotb/binary.py]: Aha yeah, I was thinking it would inherit - but that wouldn't!

File: tests/test_cases/issue_608/issue_608.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
+"""Test case for BinaryValue behaviour documented in issue #608
```
Comment (mciepluc) [inline tests/test_cases/issue_608/issue_608.py]: This test was also supposed to test other representations - 2's complement and signed magnitude. Is there a reason that this part has been removed? Here is the original test: https://github.com/cocotb/cocotb/pull/706/files#diff-762fb8676ef2834a87740564e711161d

**Comment by jurevreca12** [comment]
Is someone still working on this issue?

**Comment by eric-wieser** [comment]
This might be a cocotb 2.0 style change - it's really hard to fix this without breaking everyone that is working around the old behavior.

**Comment by jurevreca12** [comment]
Is there an approximate timeline on when to expect cocotb 2.0?
